[Git `master`] `-Werror` is great for CI, but not for a CMake default

Hi!

I ran into rightful compile errors from `-Werror` at…

https://github.com/google/libprotobuf-mutator/blob/5595eff26a93c91d13e8c073bd0de2b9f96d7693/CMakeLists.txt#L83

…today that was introduced as part of commit a3f0ab8c11bdb27d3ddf065e4f58494822eb7b74 in 2017.

While `-Werror` is great to have in CI, it is a troublesome default, in particular with compilers getting stricter by the minute (arguably for good). Could you move the `-Werror` to a CI-only place of the code base?

Thanks and best, Sebastian

CC @vitalybuka

PS: If of interest: it's GCC 13 and protobuf 29.2.0 that I see compile errors from `-Werror` with, precisely `-Werror=deprecated-declarations`:

```console
# git rev-parse HEAD
eb634efbedbf28dd6627ee49ea04e82ae144fb8d

# /usr/bin/c++ -dumpversion
13

# pkg-config protobuf --modversion
29.2.0

# make VERBOSE=1
/usr/bin/cmake -S/tmp/tmp.Ua56J9LBs4/libprotobuf-mutator -B/tmp/tmp.Ua56J9LBs4/libprotobuf-mutator --check-build-system CMakeFiles/Makefile.cmake 0
/usr/bin/cmake -E cmake_progress_start /tmp/tmp.Ua56J9LBs4/libprotobuf-mutator/CMakeFiles /tmp/tmp.Ua56J9LBs4/libprotobuf-mutator//CMakeFiles/progress.marks
make  -f CMakeFiles/Makefile2 all
make[1]: Entering directory '/tmp/tmp.Ua56J9LBs4/libprotobuf-mutator'
make  -f src/CMakeFiles/protobuf-mutator.dir/build.make src/CMakeFiles/protobuf-mutator.dir/depend
make[2]: Entering directory '/tmp/tmp.Ua56J9LBs4/libprotobuf-mutator'
cd /tmp/tmp.Ua56J9LBs4/libprotobuf-mutator && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /tmp/tmp.Ua56J9LBs4/libprotobuf-mutator /tmp/tmp.Ua56J9LBs4/libprotobuf-mutator/src /tmp/tmp.Ua56J9LBs4/libprotobuf-mutator /tmp/tmp.Ua56J9LBs4/libprotobuf-mutator/src /tmp/tmp.Ua56J9LBs4/libprotobuf-mutator/src/CMakeFiles/protobuf-mutator.dir/DependInfo.cmake "--color="
Dependencies file "src/CMakeFiles/protobuf-mutator.dir/binary_format.cc.o.d" is newer than depends file "/tmp/tmp.Ua56J9LBs4/libprotobuf-mutator/src/CMakeFiles/protobuf-mutator.dir/compiler_depend.internal".
Dependencies file "src/CMakeFiles/protobuf-mutator.dir/mutator.cc.o.d" is newer than depends file "/tmp/tmp.Ua56J9LBs4/libprotobuf-mutator/src/CMakeFiles/protobuf-mutator.dir/compiler_depend.internal".
Consolidate compiler generated dependencies of target protobuf-mutator
make[2]: Leaving directory '/tmp/tmp.Ua56J9LBs4/libprotobuf-mutator'
make  -f src/CMakeFiles/protobuf-mutator.dir/build.make src/CMakeFiles/protobuf-mutator.dir/build
make[2]: Entering directory '/tmp/tmp.Ua56J9LBs4/libprotobuf-mutator'
[ 12%] Building CXX object src/CMakeFiles/protobuf-mutator.dir/mutator.cc.o
cd /tmp/tmp.Ua56J9LBs4/libprotobuf-mutator/src && /usr/bin/c++  -I/tmp/tmp.Ua56J9LBs4/libprotobuf-mutator -fno-exceptions -Werror -Wall -MD -MT src/CMakeFiles/protobuf-mutator.dir/mutator.cc.o -MF CMakeFiles/protobuf-mutator.dir/mutator.cc.o.d -o CMakeFiles/protobuf-mutator.dir/mutator.cc.o -c /tmp/tmp.Ua56J9LBs4/libprotobuf-mutator/src/mutator.cc
/tmp/tmp.Ua56J9LBs4/libprotobuf-mutator/src/mutator.cc: In function ‘const google::protobuf::Any* protobuf_mutator::{anonymous}::CastToAny(const google::protobuf::Message*)’:
/tmp/tmp.Ua56J9LBs4/libprotobuf-mutator/src/mutator.cc:384:50: error: ‘const T* google::protobuf::DownCastToGenerated(const MessageLite*) [with T = Any]’ is deprecated [-Werror=deprecated-declarations]
  384 |              ? protobuf::DownCastToGenerated<Any>(message)
      |                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~
In file included from /usr/include/google/protobuf/generated_message_tctable_decl.h:24,
                 from /usr/include/google/protobuf/any.pb.h:23,
                 from /tmp/tmp.Ua56J9LBs4/libprotobuf-mutator/port/protobuf.h:20,
                 from /tmp/tmp.Ua56J9LBs4/libprotobuf-mutator/src/mutator.h:28,
                 from /tmp/tmp.Ua56J9LBs4/libprotobuf-mutator/src/mutator.cc:15:
/usr/include/google/protobuf/message_lite.h:1431:10: note: declared here
 1431 | const T* DownCastToGenerated(const MessageLite* from) {
      |          ^~~~~~~~~~~~~~~~~~~
/tmp/tmp.Ua56J9LBs4/libprotobuf-mutator/src/mutator.cc: In function ‘google::protobuf::Any* protobuf_mutator::{anonymous}::CastToAny(google::protobuf::Message*)’:
/tmp/tmp.Ua56J9LBs4/libprotobuf-mutator/src/mutator.cc:390:50: error: ‘T* google::protobuf::DownCastToGenerated(MessageLite*) [with T = Any]’ is deprecated [-Werror=deprecated-declarations]
  390 |              ? protobuf::DownCastToGenerated<Any>(message)
      |                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~
/usr/include/google/protobuf/message_lite.h:1437:4: note: declared here
 1437 | T* DownCastToGenerated(MessageLite* from) {
      |    ^~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
make[2]: *** [src/CMakeFiles/protobuf-mutator.dir/build.make:90: src/CMakeFiles/protobuf-mutator.dir/mutator.cc.o] Error 1
make[2]: Leaving directory '/tmp/tmp.Ua56J9LBs4/libprotobuf-mutator'
make[1]: *** [CMakeFiles/Makefile2:115: src/CMakeFiles/protobuf-mutator.dir/all] Error 2
make[1]: Leaving directory '/tmp/tmp.Ua56J9LBs4/libprotobuf-mutator'
make: *** [Makefile:136: all] Error 2
```

--- a/.github/workflows/cmake-multi-platform.yml
+++ b/.github/workflows/cmake-multi-platform.yml
@@ -50,7 +50,9 @@ jobs:
       run: >
         cmake -B ${{ steps.strings.outputs.build-output-dir }}
         -DCMAKE_CXX_COMPILER=${{ matrix.cpp_compiler }}
+        -DCMAKE_CXX_FLAGS='-Werror'
         -DCMAKE_C_COMPILER=${{ matrix.c_compiler }}
+        -DCMAKE_C_FLAGS='-Werror'
         -DCMAKE_BUILD_TYPE=${{ matrix.build_type }}
         -DLIB_PROTO_MUTATOR_DOWNLOAD_PROTOBUF=ON
         -GNinja

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -80,7 +80,7 @@ unset(CMAKE_REQUIRED_FLAGS)
 check_cxx_compiler_flag(-Wstring-conversion LIB_PROTO_MUTATOR_HAS_WSTRING_CONVERSION)
 
 if (NOT MSVC)
-  set(EXTRA_FLAGS "-fno-exceptions -Werror -Wall")
+  set(EXTRA_FLAGS "-fno-exceptions -Wall")
   if (LIB_PROTO_MUTATOR_HAS_WSTRING_CONVERSION)
     set(EXTRA_FLAGS "${EXTRA_FLAGS} -Wstring-conversion")
   endif()

